### ISSUE TITLE
R 3.4.1 warnings 

### ISSUE THREAD

**Issue by mikldk** [mikldk/ryacas#12]
I am trying to use Ryacas 0.3-1 with R 3.4.1, and I get these warnings:

```
Warning messages:
1: In structure(x$children, class = "XMLNodeList") :
    Calling 'structure(NULL, *)' is deprecated, as NULL cannot have
attributes.
    Consider 'structure(list(), *)' instead. 
```

The warnings arise in the orphaned `XML` package that is called from `OpenMath2R`. A patch suggestion using the `xml2` package instead is below:

```
diff --git a/R/OpenMath2R.R b/R/OpenMath2R.R
index 8738349..957ce60 100644
--- a/R/OpenMath2R.R
+++ b/R/OpenMath2R.R
@@ -1,35 +1,35 @@
-
 OpenMath2R <- function(x) {
  out <- c()
  recurse <- function( x ) {
-	if ("name" %in% names(xmlAttrs(x))) {
-		out <<- c(out, trans(xmlAttrs(x)[["name"]], from="OM", to="R"), " ")
+	if ("name" %in% names(xml2::xml_attrs(x))) {
+		out <<- c(out, trans(xml2::xml_attr(x, "name"), from="OM", to="R"), " ")
 	}
-	if (xmlName(x) == "text") out <<- c(out, xmlValue(x), " ")
-	if (xmlName(x) == "OMF") out <<- c(out, xmlAttrs(x)[["dec"]], " ")
-	if (xmlName(x) == "OMS") {
-	if (xmlAttrs(x)[["cd"]] == "logic1" && "name" %in% names(xmlAttrs(x))
-		&& xmlAttrs(x)[["name"]] %in% c("true", "false")) {}
-	   else if ((xmlAttrs(x)[["cd"]] != "nums1") ||
-				(xmlAttrs(x)[["name"]] == "rational")) 
-			out <<- c(out, xmlValue(x), "(")
+	if (xml2::xml_name(x) == "text") out <<- c(out, xml2::xml_text(x), " ")
+	if (xml2::xml_name(x) == "OMF") out <<- c(out, xml2::xml_attr(x, "dec"), " ")
+	if (xml2::xml_name(x) == "OMI") out <<- c(out, xml2::xml_text(x), " ")
+	if (xml2::xml_name(x) == "OMS") {
+	if (xml2::xml_attr(x, "cd") == "logic1" && "name" %in% names(xml2::xml_attrs(x))
+		&& xml2::xml_attr(x, "name") %in% c("true", "false")) {}
+	   else if ((xml2::xml_attr(x, "cd") != "nums1") ||
+				(xml2::xml_attr(x, "name") == "rational")) 
+			out <<- c(out, xml2::xml_text(x), "(")
 	}
-	# if (xmlName(x) == "OMS") out <<- c(out, "(")
-	if (xmlName(x) == "OMSTR") {
-	# out <<- c(out, sQuote(gsub("'", "\\\\'", xmlValue(x))))
-	out <<- c(out, paste("'", gsub("'", "\\\\'", xmlValue(x)), "'", sep=""))
-	} else if ( length( xmlChildren(x) ) > 0 )
-		for( i in seq( along = xmlChildren(x) ) ) {
-			Recall( x[[i]] )
-			if (i > 1 && i < length(xmlChildren(x))) 
+	# if (xml2::xml_name(x) == "OMS") out <<- c(out, "(")
+	if (xml2::xml_name(x) == "OMSTR") {
+	# out <<- c(out, sQuote(gsub("'", "\\\\'", xml2::xml_text(x))))
+	out <<- c(out, paste("'", gsub("'", "\\\\'", xml2::xml_text(x)), "'", sep=""))
+	} else if ( length( xml2::xml_children(x) ) > 0 )
+		for( i in seq( along = xml2::xml_children(x) ) ) {
+			Recall( xml2::xml_children(x)[[i]] )
+			if (i > 1 && i < length(xml2::xml_children(x)))
 				out <<- c(out, ",")
 		}
-	# if (xmlName(x) == "OMA" || xmlName(x) == "OMBIND") out <<- c(out, xmlValue(x), ")")
-	if (xmlName(x) == "OMA" || xmlName(x) == "OMBIND") out <<- c(out, ")")
+	# if (xml2::xml_name(x) == "OMA" || xml2::xml_name(x) == "OMBIND") out <<- c(out, xml2::xml_text(x), ")")
+	if (xml2::xml_name(x) == "OMA" || xml2::xml_name(x) == "OMBIND") out <<- c(out, ")")
  }
  x <- paste(x, "\n", collapse = "")
- x <- xmlTreeParse(x, asText = TRUE)
- x <- xmlRoot(x)
+ x <- xml2::read_xml(x)
+ x <- xml2::xml_root(x)
  recurse(x)
  paste(out, collapse = "")
 }
```

**Comment by HaoLi111** [comment]
Compare the 2 ways(1st gives warnings whereas 2nd one does not)
> structure(NULL, class = "XMLNodeList")
list()
attr(,"class")
[1] "XMLNodeList"
Warning message:
In structure(NULL, class = "XMLNodeList") :
  Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.
  Consider 'structure(list(), *)' instead.
> structure(list(),class = "XMLNodeList")
list()
attr(,"class")
[1] "XMLNodeList"
> 

So if we do

ifelse(is.null(x$children),structure(list(),class = "XMLNodeList"),structure(x$children, class = "XMLNodeList"))

to substitute the original line,
but where are they? Maybe they are functions that are called by others and then called by Ryacas?

**Comment by haakonstorm** [comment]
R 3.5.1 on macOS High Sierra here
yacas 1.6.1 installed from mac binary
Ryacas installed via devtools::install_github

These warning messages appear consistently:
Warning messages:
1: In structure(x$children, class = "XMLNodeList") :
  Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.
  Consider 'structure(list(), *)' instead.
2: In structure(x$children, class = "XMLNodeList") :
  Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.
  Consider 'structure(list(), *)' instead.
3: In structure(x$children, class = "XMLNodeList") :
  Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.
  Consider 'structure(list(), *)' instead.

**Comment by ggrothendieck** [comment]
> R 3.5.1 on macOS High Sierra here
> yacas 1.6.1 installed from mac binary
> Ryacas installed via devtools::install_github
> 
> These warning messages appear consistently:
> Warning messages:
> 1: In structure(x$children, class = "XMLNodeList") :
> Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.
> Consider 'structure(list(), *)' instead.
> 2: In structure(x$children, class = "XMLNodeList") :
> Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.
> Consider 'structure(list(), *)' instead.
> 3: In structure(x$children, class = "XMLNodeList") :
> Calling 'structure(NULL, *)' is deprecated, as NULL cannot have attributes.
> Consider 'structure(list(), *)' instead.

I have had Ryacas removed from CRAN until someone is found to take over maintenance.  if you are interested let me know.

**Comment by haakonstorm** [comment]
Aha, get it, I'm not skilled enough in these areas to contribute. Anyways, thanks
a lot of Ryacas. It's great!
Storm

**Comment by mikldk** [comment]
@ggrothendieck I am willing to take over the maintenance and see if I can get the package updated to not give warnings. How is it done in practise?

**Comment by HaoLi111** [comment]
**1. Download and built with ctrl shift B in Rstudio.--No problem so far
> Factor(expression(x^2-1))
expression((x + 1) * (x - 1))
#good work
2. devtools::install_github('mikldk/ryacas') -- Fails as Rtools is unavailable for R 3.5.1**

**I am using**

    **_R version 3.5.1 (2018-07-02) -- "Feather Spray"
    Copyright (C) 2018 The R Foundation for Statistical Computing
    Platform: x86_64-w64-mingw32/x64 (64-bit)_**

**and get following:(not the problem of the package itself, maybe a problem of devtools?  as I do have Rtools 3.5.0.2, Rtools 3.5 is not frozen so far...)**
[A link to Rtools](https://cran.r-project.org/bin/windows/Rtools/)
     > devtools::install_github('mikldk/ryacas')
Downloading GitHub repo mikldk/ryacas@master
from URL https://api.github.com/repos/mikldk/ryacas/zipball/master
WARNING: Rtools is required to build R packages, but no version of Rtools compatible with R 3.5.1 was found. (Only the following incompatible version(s) of Rtools were found:3.5)

Please download and install the appropriate version of Rtools from http://cran.r-project.org/bin/windows/Rtools/.
Installing Ryacas
"C:/PROGRA~1/R/R-35~1.1/bin/x64/R" --no-site-file --no-environ --no-save --no-restore --quiet CMD INSTALL  \
  "C:/Users/DELL/AppData/Local/Temp/Rtmp6P4tlG/devtools2b603d060ec/mikldk-ryacas-0a8bdb9" --library="C:/Users/DELL/Documents/R/win-library/3.5" --install-tests 

* installing *source* package 'Ryacas' ...
** libs

*** arch - i386
Warning in system(cmd) : 'make' not found
ERROR: compilation failed for package 'Ryacas'
* removing 'C:/Users/DELL/Documents/R/win-library/3.5/Ryacas'
* restoring previous 'C:/Users/DELL/Documents/R/win-library/3.5/Ryacas'
Warning in file.copy(lp, dirname(pkgdir), recursive = TRUE, copy.date = TRUE) :
  拷贝C:\Users\DELL\Documents\R\win-library\3.5\00LOCK-mikldk-ryacas-0a8bdb9\Ryacas\libs\x64\Ryacas.dll到C:\Users\DELL\Documents\R\win-library\3.5\Ryacas\libs\x64\Ryacas.dll时出了问题：Permission denied 
In R CMD INSTALL
Installation failed: Command failed (1)

**As for Rtools**

> install.packages('Rtools')
Installing package into ‘C:/Users/DELL/Documents/R/win-library/3.5’
(as ‘lib’ is unspecified)
Warning in install.packages :
  package ‘Rtools’ is not available (for R version 3.5.1)

**Comment by mikldk** [comment]
@HaoLi111 Thanks. It looks like to be a problem with your combination of software versions.

**Comment by haakonstorm** [comment]
Hey! Built and run beautifully here! Thanks a bunch :)

**Comment by mikldk** [comment]
Perfect, thanks.
Fixed in e31954259b4ee9e23b0566566b94a9cad7e32ab8
